lc2fen/infer_pieces.py

```diff
@@ -164,26 +168,124 @@ def infer_chess_pieces(pieces_probs, a1_pos, previous_fen=None):
         If it is not None, improves piece inference.
     :return: A list of the inferred chess pieces in FEN notation order.
     """
+    if previous_fen is not None:
+        previous_board = chess.Board(previous_fen)
+        previous_list = board_to_list(fen_to_board(previous_fen))
     pieces_probs = board_to_list(list_to_board(pieces_probs, a1_pos))
 
     # None represents that no piece is set in that position yet
     out_preds = [None] * 64
 
     final_move_sq = -1
-    possible_pieces = []
-    if previous_fen is not None:
+    if previous_fen is not None:  # Perform move detection
         changed_squares_idx = changed_squares(previous_fen, pieces_probs)
         move = inferred_move(previous_fen, pieces_probs, changed_squares_idx)
-        if move is not None:
+        if (
+            move is not None
+        ):  # A move has been successfully detected so the FEN will be concluded immediately
             initial_sq, final_move_sq, action = move
-            possible_pieces = inferred_pieces_from_move(
-                initial_sq, final_move_sq, action
-            )
+            initial_coordinates = FILES[initial_sq % 8] + RANKS[initial_sq // 8]
+            final_coordinates = FILES[final_move_sq % 8] + RANKS[final_move_sq // 8]
+            move_UCI = initial_coordinates + final_coordinates
+            if action.startswith("white"):
+                previous_board.turn = chess.WHITE
+            else:
+                previous_board.turn = chess.BLACK
+            if (
+                previous_list[initial_sq] == "P" and initial_coordinates[1] == "7"
+            ):  # White promotes (and we have to figure out the promoted piece)
```

Comment:
Could you fuse this and the black promotion into a separate auxiliary function to clarify the code? If not, this function gets too long. It could be similar to the `__max_piece` function.

Reply:
Done!